### PR TITLE
feat(backend): Make Redis connection Sync + Use Redis as Distributed Lock

### PR DESCRIPTION
**Pull request by majdyz**
### Background

Main issue:
We rely on the idea of rest_api being a singleton service to maintain our distributed lock. This will cause issues when the number of services increases.

Side issue:
The Redis operation that is used in our system has nothing to do with our API request:
* For read & write to queue done by a background process.
* (Added) To acquire a mutually exclusive lock.

Using Redis on asyncio will add unnecessary overhead and overscribe the event loop shared by Prisma queries.

### Changes 🏗️
* Migrate distributed lock from rest_api to Redis.
* Switch Redis to synchronous client.
* Cleanup on connection retry logic on db.py and redis.py (newly introduced).
* Remove InMemoryEventQueue and make integration test use Redis queue.

Only for the new autogpt platform, currently in autogpt_platform/

- Create from scratch and execute an agent with at least 3 blocks
- Import an agent from file upload, and confirm it executes correctly
- Upload agent to marketplace
- Import an agent from marketplace and confirm it executes correctly
- Edit an agent from monitor, and confirm it executes correctly

### CODE REVIEW

File: autogpt_platform/backend/backend/data/db.py [hunk 1]
```diff
@@ -1,4 +1,3 @@
-import asyncio
```
Comment (majdyz) [inline autogpt_platform/backend/backend/data/db.py:1]: No change here, just connection code tidying.
I decoupled the retry logic and moved it to retry.py.

File: autogpt_platform/backend/backend/data/queue.py [hunk 2]
```diff
@@ -18,60 +16,46 @@ def default(self, o):
         return super().default(o)
 
 
-class AsyncEventQueue(ABC):
+class AbstractEventQueue(ABC):
```
Comment (majdyz) [inline autogpt_platform/backend/backend/data/queue.py:19]: I made the EventQueue sync, all the callers are blocking anyway.

File: autogpt_platform/backend/backend/data/redis.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import logging
```
Comment (majdyz) [inline autogpt_platform/backend/backend/data/redis.py:1]: similar to db.py. This forces a single Redis connection per process.
The redis connection used here is sync, fast API (if it even need redis) will do it separately anyway.

File: autogpt_platform/backend/backend/executor/manager.py [hunk 4]
```diff
@@ -17,7 +17,7 @@
 if TYPE_CHECKING:
     from backend.server.rest_api import AgentServer
 
-from backend.data import db
+from backend.data import db, redis
```
Comment (majdyz) [inline autogpt_platform/backend/backend/executor/manager.py:20]: Basically, this is the main changes, locking use redis, and remove the rest_server.lock

File: autogpt_platform/backend/backend/server/rest_api.py [hunk 5]
```diff
@@ -32,24 +31,23 @@
...
-    use_redis = True
+    use_queue = True
     _test_dependency_overrides = {}
     _user_credit_model = get_user_credit_model()
 
-    def __init__(self, event_queue: AsyncEventQueue | None = None):
+    def __init__(self):
         super().__init__(port=Config().agent_server_port)
-        self.event_queue = event_queue or AsyncRedisEventQueue()
+        self.event_queue = RedisEventQueue()
```
Comment (majdyz) [inline autogpt_platform/backend/backend/server/rest_api.py:40]: Since we are already forcing Redis, there is no added value of mocking the queue in our integration test.
OTOH, we need some test coverage on this.

File: autogpt_platform/backend/backend/util/service.py [hunk 6]
```diff
@@ -45,9 +45,9 @@ def wrapper(*args, **kwargs):
 
 class AppService(AppProcess):
     shared_event_loop: asyncio.AbstractEventLoop
-    event_queue: AsyncEventQueue = AsyncRedisEventQueue()
+    event_queue: AbstractEventQueue = RedisEventQueue()
     use_db: bool = False
-    use_redis: bool = False
+    use_queue: bool = False
```
Comment (majdyz) [inline autogpt_platform/backend/backend/util/service.py:50]: Minor renaming, since we use redis for something else now aside from the queue

File: .github/workflows/platform-backend-ci.yml [hunk 7]
```diff
@@ -32,6 +32,13 @@ jobs:
         python-version: ["3.10"]
     runs-on: ubuntu-latest
 
+    services:
```
Comment (majdyz) [inline .github/workflows/platform-backend-ci.yml:35]: Somehow this change makes the CI not kick in (also not failed either).
cc @Pwuts any clue?

Comment (kcze) [reply .github/workflows/platform-backend-ci.yml:35]: It fails but I don't understand what happened:
https://github.com/Significant-Gravitas/AutoGPT/actions/runs/11063086188?pr=8197

Comment (majdyz) [reply .github/workflows/platform-backend-ci.yml:35]: Okay it's working now.. Thanks!!

File: autogpt_platform/backend/backend/util/retry.py [hunk 8]
```diff
@@ -1,7 +1,44 @@
...
+            await asyncio.sleep(2**call_count)
+            return await logged_retry(
+                func=func,
+                resource_name=resource_name,
+                action_name=action_name,
+                call_count=call_count + 1,
+                max_retry=max_retry,
+            )
+        else:
+            raise e
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/util/retry.py]: Why not just use `tenacity`?

File: autogpt_platform/backend/backend/executor/manager.py [hunk 9]
```diff
@@ -453,6 +455,7 @@ def on_node_executor_stop(cls):
             return  # already shutting down
 
         logger.info(f"[on_node_executor_stop {cls.pid}] ⏳ Disconnecting DB...")
+        redis.disconnect()
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/executor/manager.py]: please move this below the `db.disconnect()`, for 2 reasons:
- it's now directly after a log statement that says "Disconnecting DB"
- startup/shutdown sequences should be each other's reverse

File: autogpt_platform/backend/backend/util/retry.py (hunk 8 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/util/retry.py]: Updated

File: autogpt_platform/backend/backend/executor/manager.py (hunk 9 above)
Comment (majdyz) [reply autogpt_platform/backend/backend/executor/manager.py]: Updated

File: autogpt_platform/backend/backend/data/block.py [hunk 10]
```diff
@@ -286,7 +286,9 @@ def get_blocks() -> dict[str, Block]:
 
 async def initialize_blocks() -> None:
     for block in get_blocks().values():
-        existing_block = await AgentBlock.prisma().find_unique(where={"id": block.id})
+        existing_block = await AgentBlock.prisma().find_first(
+            where={"OR": [{"id": block.id}, {"name": block.name}]}
+        )
```
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/block.py:291]: If you have run the migration this shouldn't be necessary

File: autogpt_platform/backend/backend/data/block.py
Comment (Pwuts) [inline autogpt_platform/backend/backend/data/block.py:1]: This is unrelated and shouldn't be necessary if you just run the migration added in #8229

Comment (majdyz) [reply autogpt_platform/backend/backend/data/block.py:1]: I'm gonna keep this to be in the safe side. It's very easy to change stuff and forgot adding a migration. Also if the contributors are not aware of it, let alone the other users.

Do you want me to decouple this change from this PR?

**Comment by aarushik93** [comment]
testing this PR today, couldn't get to it yesterday

File: autogpt_platform/backend/backend/data/block.py
Comment (Pwuts) [reply autogpt_platform/backend/backend/data/block.py:1]: I'd really like to know why you got those errors despite the migration which should prevent it. Probably hard to find out now, but I really don't see what could be going wrong here. It worked perfectly for me (locally).

> Do you want me to decouple this change from this PR?

preferably yes

Comment (majdyz) [reply autogpt_platform/backend/backend/data/block.py:1]: > I'd really like to know why you got those errors despite the migration which should prevent it. Probably hard to find out now, but I really don't see what could be going wrong here. It worked perfectly for me (locally).

This is my educated guess is:
* I pulled master 
* Your migration is there
* I checkout other branch
* Your migration is gone, but the directory is still there
* Since the docker app auto-run your migration, the state we are in is you have a messed up migration state.

I don't have time to confirm this though unfortunately.

> preferably yes

Moved here, appreciate your review on that too:
https://github.com/Significant-Gravitas/AutoGPT/pull/8249

**Review by aarushik93 — APPROVED** [review]
